crawler.py

```diff
@@ -111,37 +115,60 @@
 
 unsaved_jobs = list()
 unsaved_jobs_lock = Lock()
+
+
+def get_job_completion_callback(logger: logging.Logger,
+                                unsaved_jobs_lock: Lock,
+                                job_queue: rediswq.RedisWQ,
+                                job: bytes) -> Callable[[], None]:
+    def callback():
+        with unsaved_jobs_lock:
+            logger.info("Job %r is done", job)
+            job_queue.complete(job)
+            unsaved_jobs.remove(job)
+    return callback
+
+
+no_job_since = None
 # Crawl sites specified in job queue until empty
 while not job_queue.empty():
     job_queue.check_expired_leases()
-    with(unsaved_jobs_lock):
+    with unsaved_jobs_lock:
+        manager.logger.debug("Currently unfinished jobs are: %s", unsaved_jobs)
         for unsaved_job in unsaved_jobs:
             if not job_queue.renew_lease(unsaved_job,
-                                         2 * (TIMEOUT + DWELL_TIME + 30)):
+                                         EXTENDED_LEASE_TIME):
                 manager.logger.error("Unsaved job: %s timed out", unsaved_job)
 
     job = job_queue.lease(
         lease_secs=TIMEOUT + DWELL_TIME + 30, block=True, timeout=5
     )
     if job is None:
-        manager.logger.info("Waiting for work")
+        if no_job_since is None:
```

Comment:
This is needed since every job doesn't get saved until a new one started, so the last job of each browser stays unfinished until `close` gets called, however `close` can't get called as there are still jobs in `:processing` so the job_queue isn`t empty.
I attempt to resolve this deadlock by waiting for the extended lease time, thus ensuring all unfinished jobs are being held by other instances or ourselves and thus implying that the job didn't fail but everybody is just waiting on each other, each instance then chooses to terminate itself to resolve this deadlock and since they all call `close` everything gets saved out.

Reply:
@englehardt I can write up a longer form piece on this if you want me to.
This feels very hacky and I don't like it, but it's the best I came up with.